### ISSUE TITLE
Refactoring processAmberForceField.py

### ISSUE THREAD

**Issue by jchodera** [pandegroup/openmm#1184]
We would like to refactor [`processAmberForceField.py`](https://github.com/pandegroup/openmm/blob/master/devtools/forcefield-scripts/processAmberForceField.py) to accomplish the following goals:
* Be more readable and maintainable to other developers
* Produce AMBER `ffxml` files that use new `ffxml` extensions (#1168) such as the per-atom typing scheme (#1182) and the name-based `<Bond>` specification schemes (proposed in #1168)
* Correctly handle `prep` files (see #1181)
* Allow output of parameters and residue definitions into different `ffxml` files (if supported by `ForceField`) so that we can combine `gaff.xml`, small molecule definitions, and `gbsa` parameters in a "mix-and-match" way.
* Convert small molecules in addition to polymeric residues, and allow these molecules to be loaded from a separate `ffxml` file by avoiding type name definition collisions
* Add automated nosetests to ensure that conversion of polymeric residues and small molecules gives correct parameters
* Enable the addition of an external automation harness to be able to automate the conversion of AMBER forcefields (which @rafwiewiora is working on in #1134) and inclusion of references for provenance data.

Much of this effort was actually already undertaken by @kyleabeauchamp and @juliebehr in what is now in [openmoltools](https://github.com/choderalab/openmoltools/blob/master/openmoltools/amber_parser.py), where readability has been significantly improved and small molecules can be converted.  Testing frameworks have also been written, including [test utilities](https://github.com/choderalab/openmoltools/blob/master/openmoltools/utils.py#L260-L287) and [multiple](https://github.com/choderalab/openmoltools/blob/master/openmoltools/tests/test_drugs.py) [nosetests](https://github.com/choderalab/openmoltools/blob/master/openmoltools/tests/test_common_molecules.py).

I'd like to start with the [refactored class-based version of the script](https://github.com/choderalab/openmoltools/blob/master/openmoltools/amber_parser.py) and clean this up to the point where it accomplishes these goals.

**Comment by jchodera** [comment]
I should note that this refactored version only has some of the capabilities of the full script, but the proposed approach is to continue to convert the other capabilities in a similar manner.

I would eliminate the use of numeric atom types, instead only supporting the use of named AMBER atom types in the new version.

**Comment by jchodera** [comment]
Some additional thoughts here:
* If we are adopting a new class-based infrastructure, we might as well make the interface publicly accessible through `simtk.openmm.app` to let users call the converter externally and produce `ffxml` files from AMBER input files (polymers, carbohydrates, cofactors, or small molecules).  This could be generalized to a `ForceFieldConverter` base class with `AmberForceFieldConverter` being a subclass.
* I realize that [`parmed`](https://github.com/parmed/parmed) from @swails may have some of this capability already, but I'm not sure how much functionality for converting forcefield *definition* files there is, rather than converting individual systems.

**Comment by swails** [comment]
> we might as well make the interface publicly accessible through simtk.openmm.app to let users call the converter externally and produce ffxml files from AMBER input files

My personal thought is that this is rather orthogonal to what `simtk.openmm.app` is and does, and that another namespace might be more appropriate (or maybe have it go inside `simtk.openmm.app.internal` and have that functionality be available as a backend to some tools in `app` that may wind up using it).  It can always be moved *into* `app` later if it's clearly the way to go (but it can never really be moved out, or changed, otherwise).

> I realize that parmed from @swails may have some of this capability already, but I'm not sure how much functionality for converting forcefield definition files there is, rather than converting individual systems.

ParmEd can read Amber parameter and OFF files, as well as leaprc files, and generate residue definitions.  It can also read CHARMM prm and RTF files (and STR) files, including residue definitions.  It does not yet write out OpenMM XML files, but all of the "hard" stuff is already there.  And the data structures would make it trivial to write out the OpenMM variants.  (It also does GROMACS parameter files, although not the residue definitions yet).  My *eventual* goal here is to have ParmEd replace leap, so I'm adding components that are related to things I need to do for present applications in a way to make my long-term goal incrementally easier.

The files it does not yet process are Amber PREP files (since we are actively trying to eliminate that format in favor of OFF and Mol2 files -- both of which ParmEd reads fairly robustly).  Although if that was a severe limitation, PREP support could be added as well.

If you wanted to go through ParmEd for this functionality, then I could make it a priority to add.  If not, then I'd wait on it.

**Comment by jchodera** [comment]
Having this functionality accessible through an API has numerous advantages:
* There is currently no easy way for users to import other forcefields that may be in Amber format
* There is an Amber parameter repository of prep/off/in files for cofactors, ions, etc that would be useful to facilitate conversion from so users can use these if they want to use some new parameters that are deposited
* There is still no real way to convert small molecules, which is a behavior I would consider essential. gaff2xml (now in openmoltools) essentially reimplemented a modified subset  of the parameter conversion script functionality and refactored it to provide a user-facing API, which is very useful. But we probably want to talk about how to bring some of that functionality into OpenMM soon
- It is easier to code up testing harnesses

Regarding parmed:  If parmed already has much of this functionality, and will likely be supported moving forward, maybe we want to switch to that for forcefield conversion?

Regarding leap replacement: Are there plans to support small molecule conversion in ParmEd? Can you already combine and set up systems (eg protein, small molecule, add solvent) in the same way you can with leap, or is that envisioned future functionality?

**Comment by swails** [comment]
> Having this functionality accessible through an API has numerous advantages:

I think a question worth answering here is *where* should that API go (not *should there be* an API)?

> There is an Amber parameter repository of prep/off/in files for cofactors, ions, etc that would be useful to facilitate conversion from so users can use these if they want to use some new parameters that are deposited

Sure.  But how should this be handled?  Should this be something *we* do and provide files for so users themselves don't have to?  That would get my vote.

> There is still no real way to convert small molecules, which is a behavior I would consider essential. gaff2xml (now in openmoltools) essentially reimplemented a modified subset  of the parameter conversion script functionality and refactored it to provide a user-facing API, which is very useful. But we probably want to talk about how to bring some of that functionality into OpenMM soon

Agreed, and right now users have to go through CGenFF or antechamber to do so.  It seems to me that the workflow for doing custom residues in *any* force field (CHARMM, AMOEBA, and AMBER) is to prepare the whole thing in the other program and use the file parsers to go to OpenMM.

How should a small-molecule prep utility work in OpenMM?  openmoltools is certainly one approach, but that is limited entirely to gaff and Amber's workflow.  You could presumably do the same kind of wrapping around CGenFF.  But ideally what I'd like to see in a user-facing API (i.e., that which we put in `simtk.openmm.app`) is "here is a small-molecule residue, give me parameters for force field X", then `ForceField` goes off and does it (and provides an option to save an XML for future use).  In which case, all of these "conversion utilities" really belong in `simtk.openmm.app.internal` so that `ForceField` can make use of them to extend its reach to small molecules.

My vision of `simtk.openmm.app` is that it provides parsers for other programs parameter-topology files as well as a way to do it [Omnia-style](https://www.youtube.com/watch?v=9bZkp7q19f0).  A solution that tethers to specific parameter file formats from other packages in order to try and provide a *general* solution to small-molecule parametrization within the ForceField approach seems wrong to me.  I think it should be "provide a prmtop or ask ForceField for a particular Amber FF".  Not "ask ForceField for a particular Amber FF and piece together these other Amber intermediate files to provide parameters for a small molecule to complete the picture".  Let's wrap up that last part as an implementation detail in ForceField to handle small molecules until we can come up with something better.  The same way that all of the Amber GB models are in `app.internal`, but are still available through `app` by asking for those models from the various file parsers.

> Regarding parmed:  If parmed already has much of this functionality, and will likely be supported moving forward, maybe we want to switch to that for forcefield conversion?

I'd be on board, but then such functionality would either not be part of OpenMM proper, or would only be accessible within OpenMM for users that also installed ParmEd (another argument to keep it in `simtk.openmm.app.internal` if the latter is chosen, IMO).

> Regarding leap replacement: Are there plans to support small molecule conversion in ParmEd?

There is some level of that already.  It creates residue templates from any of a number of file formats (CHARMM RTF, Amber OFF, Mol2, and pretty soon component CIF).  I could always wrap antechamber as well, although until I (or another ParmEd user) have a need, I'm not committing to anything.

> Can you already combine and set up systems (eg protein, small molecule, add solvent) in the same way you can with leap, or is that envisioned future functionality?

Again, it has some components that will help toward that end, but a lot of the LEaP replacement stuff is envisioned future functionality.

**Comment by jchodera** [comment]
> I think a question worth answering here is where should that API go (not should there be an API)?

Also, "who is the API intended for? developers or users?".  My feeling is that we provide tools for importing Amber `prmtop/inpcrd` files into OpenMM native formats, so providing tools for converting Amber `prep/in/off/lib` into OpenMM `app` native formats should be considered to service the same class of users.

> Sure. But how should this be handled? Should this be something we do and provide files for so users themselves don't have to? That would get my vote.

I do agree that it would be easiest if we could do this, but we can also not control when files are added to the [Amber parameter database](http://www.pharmacy.manchester.ac.uk/bryce/amber/) or [published independently](http://septlab.engin.umich.edu/multisite-ions.html).  This might happen between OpenMM releases, so we would need to provide some way to help users convert files without having to write their own converter or use undocumented and untested code.

> Agreed, and right now users have to go through CGenFF or antechamber to do so. It seems to me that the workflow for doing custom residues in any force field (CHARMM, AMOEBA, and AMBER) is to prepare the whole thing in the other program and use the file parsers to go to OpenMM.

Preparing the *whole system* in another code is not a fantastic option.  The `ffxml` and `ForceField` facility of OpenMM can easily be extended (and really seems to be intended to in philosophy) permit the mix-and-match importing of new residue definitions, but we just need a way to facilitate this through allowing the easy import of converted residues.

> How should a small-molecule prep utility work in OpenMM? openmoltools is certainly one approach, but that is limited entirely to gaff and Amber's workflow. You could presumably do the same kind of wrapping around CGenFF. But ideally what I'd like to see in a user-facing API (i.e., that which we put in simtk.openmm.app) is "here is a small-molecule residue, give me parameters for force field X", then ForceField goes off and does it (and provides an option to save an XML for future use). In which case, all of these "conversion utilities" really belong in simtk.openmm.app.internal so that ForceField can make use of them to extend its reach to small molecules.

I'd love to see this too.

One way to permit this is to have a concept of "plug-ins" for `ForceField` that tell it how to parameterize new molecules using some standard API.  We could provide some default "plug-ins" for Antechamber and CGenFF (is something available as a stand-alone tool yet?) but users could also add their own.

> My vision of simtk.openmm.app is that it provides parsers for other programs parameter-topology files as well as a way to do it Omnia-style. A solution that tethers to specific parameter file formats from other packages in order to try and provide a general solution to small-molecule parametrization within the ForceField approach seems wrong to me. I think it should be "provide a prmtop or ask ForceField for a particular Amber FF". Not "ask ForceField for a particular Amber FF and piece together these other Amber intermediate files to provide parameters for a small molecule to complete the picture".

I certainly agree that the idea of asking `ForceField` to produce parameters for an unparameterized residue using a plugin is going to be very useful, I don't think we can avoid providing a facility for converting legacy AMBER residue/forcefield definition files if we want to support things like the [Amber parameter database](http://www.pharmacy.manchester.ac.uk/bryce/amber/).

> There is some level of that already. It creates residue templates from any of a number of file formats (CHARMM RTF, Amber OFF, Mol2, and pretty soon component CIF). I could always wrap antechamber as well, although until I (or another ParmEd user) have a need, I'm not committing to anything.

This is a nice variety of formats to support!

**Comment by swails** [comment]
> Amber prmtop/inpcrd files into OpenMM native formats, so providing tools for converting Amber prep/in/off/lib into OpenMM app native formats should be considered to service the same class of users.

As someone who falls into the first class of users, I would not develop a general workflow using these tools to convert to OpenMM.  Once you have the files you need to convert to XML, you've already done the hardest part of building a system in Amber.  I'll use what I know.  And then the new/naive user has to ask -- what should my workflow be?  How far do I go in Amber before switching to OpenMM?  All the way to the prmtop?  Halfway?  Less?  More?

> but we just need a way to facilitate this through allowing the easy import of converted residues.

This is still tying an OpenMM workflow to require bits and pieces of workflows from other programs without providing a pure-OpenMM way to do things.  What this is essentially telling brand new users is "if you want to support small molecules in OpenMM, you have to learn some Amber so you can do the first steps there before translating to OpenMM."  Compared to Amber's line of "just use the tutorials to learn how to use Amber".

> I certainly agree that the idea of asking ForceField to produce parameters for an unparameterized residue using a plugin is going to be very useful, I don't think we can avoid providing a facility for converting legacy AMBER residue/forcefield definition files if we want to support things like the Amber parameter database.

I still think we want to add them somewhere *other* than `simtk.openmm.app` and present/document them away from where we talk about ForceField to avoid undue confusion (of course this could be done even if the tools were dumped into `app`).  I also think that having our own database with our own file formats is a good thing, and should eliminate the need for users to convert existing ones.  These databases grow very slowly, so I doubt "keeping up" would present any challenge.

**Comment by mpharrigan** [comment]
Not following the whole discussion, but many lab members prepare prmtops in amber and switch to openmm. I don't think those are the same people who are itching to convert forcefields

**Comment by jchodera** [comment]
How about this proposal:

### Pure OpenMM workflow
Allows users to exclusively use `Modeller` and `ForceField` to parameterize molecules with missing templates via their preferred plugin.
* User would initialize `ForceField` with specified handlers or plugins for handling missing small molecules.  We start with an `Antechamber` based plugin, but want to add more later.
* `Modeller` or `ForceField` would identify residues with missing residue templates, and use a standard workflow for creating new residue templates. 
* We may also need to figure out how to add missing hydrogens for these new templates

### Pure external workflow
We allow users to create the whole system in another program (AMBER, CHARMM) and import the whole system into OpenMM.  This is already possible.

### Conversion of forcefield definition files and user-deposited cofactor/ion definitions in online repositories
We provide a set of tools in `simtk.openmm.app.internal` for converting forcefields
* Allows us to automate testing of conversion scripts in travis
* Allows us to write scripts that use this API for converting forcefields with provenance information
* Allows power users to convert cofactors / ion definitions, but if people end up needing this functionality more often, we add a nicer interface through `simtk.openmm.app`

**Comment by jchodera** [comment]
We'd obviously need to further refine how the pure OpenMM workflow should operate, especially for polymeric residues with external bonds (especially how charges should be derived), but the code for small molecules can be brought over from `openmoltools` with redundancies removed once the forcefield conversion tools are brought into `simtk.openmm.app.internal`.

**Comment by swails** [comment]
That's more or less what I was advocating for.

**Comment by jchodera** [comment]
Great! If @peastman is onboard too, we can start refining the details.

**Comment by peastman** [comment]
Sorry for not commenting till now.  Anyway, if you're both happy with this overall design then by all means go ahead and start discussing details.

**Comment by jchodera** [comment]
Our recent conversations have been focusing more on whether we might be able to use `ParmEd` for these conversions.  It sounds like it would be straightforward to add support for writing `ffxml` files.

@peastman: How comfortable would you be with a `ParmEd`-based solution for converting `ffxml` files, or a solution that utilizes `ParmEd` as an optional dependency for this process?

**Comment by peastman** [comment]
That would be fine.

**Comment by jchodera** [comment]
We're going with a ParmEd-based scheme (#1281) instead, so refactoring of `processAmberForceField.py` is no longer relevant.  Will spin the relevant parts of this into separate issues.
